src/main/java/com/solvd/service/LoginService.java

```diff
@@ -1,19 +1,18 @@
 package com.solvd.service;
 
-import com.solvd.pages.common.DashboardPageBase;
 import com.solvd.pages.common.LoginPageBase;
+import com.solvd.pages.common.PageWithLeftMenuBase;
 import com.zebrunner.carina.utils.R;
 import com.zebrunner.carina.utils.factory.ICustomTypePageFactory;
-import org.openqa.selenium.WebDriver;
 
 public class LoginService implements ICustomTypePageFactory {
 
 
-    public DashboardPageBase successfulLogin() {
+    public PageWithLeftMenuBase successfulLogin() {
```

Comment:
I could do class like this, I know it. Something like:

public <T extends PageWithLeftMenuBase> T successfulLogin(Class<T> page) {
        LoginPageBase loginPage = initPage(getDriver(), LoginPageBase.class);
        loginPage.open();
        loginPage.logIn(R.TESTDATA.get("correct_user"), R.TESTDATA.get("correct_password"));

        return initPage(getDriver(), page);
    }
    
 but what I would pass in page as an argument? PageWithLeftMenuBase which will be repetition. Keep in mind that I can't predict which page will be opened after each test, so if I will have to pass url or page(I think page would be most suitable) still I would have to know which one it would be, which I do not know. So current solution seems for me reasonable. Since I know that each page has basic navigation, I just use it, and then redirect to site I will know it's this site and will have this implementation.

Reply:
I know what you mean. Like this way, right?

```java
public <T extends PageWithLeftMenuBase> T successfulLogin(Class<T> page) {
        LoginPageBase loginPage = initPage(getDriver(), LoginPageBase.class);
        loginPage.open();
        loginPage.logIn(R.TESTDATA.get("correct_user"), R.TESTDATA.get("correct_password"));

        return initPage(getDriver(), page);
}
```
It can't be done 'cuz, I don't know which site I will get on. So which page I should pass there, or url? If I will pass AddEmployeePageBase.class, I ain't sure I will be redirected after login to this site. So this solution is better, more optimal. Since PageWithLeftMenuBase it's a CORE, BASE of all pages after login, and it has basic functions implemented. Then from this site I can redirect to others and then I am sure I am on this page.

Reply:
Improved as we discussed on our call, I just had to add assertion after login, works good! :)

```java
public <T extends PageWithLeftMenuBase> T successfulLogin(Class<T> page) {
        LoginPageBase loginPage = initPage(getDriver(), LoginPageBase.class);
        loginPage.open();
        loginPage.logIn(R.TESTDATA.get("correct_user"), R.TESTDATA.get("correct_password"));
        PageWithLeftMenuBase pageWithLeftMenuBase = initPage(getDriver(), PageWithLeftMenuBase.class);
        pageWithLeftMenuBase.assertPageOpened();

        T targetPage = initPage(getDriver(), page);
        targetPage.open();
        targetPage.assertPageOpened();
        return  targetPage;
    }
    ```